packages/demo-protobuf/src/decorator.ts

```diff
@@ -0,0 +1,20 @@
+import { Constructor, Field, Message, TypeDecorator, util } from "protobufjs";
+
+import { Registry } from "./registry";
+
+function getTypeName(typeUrl: string): string {
+  const parts = typeUrl.split(".");
+  return parts[parts.length - 1];
+}
+
+export function CosmosMessage(registry: Registry, typeUrl: string): TypeDecorator<any> {
+  return (constructor: Constructor<Message<any>>) => {
+    const typeName = getTypeName(typeUrl);
+    const generatedType = util.decorateType(constructor, typeName);
```

Comment:
This seems to be implemented via reflection (https://github.com/protobufjs/protobuf.js/blob/v6.9.0/src/util.js#L113-L143), which is great since we doon't need the .proto or JSON implementation at all then.